network/discovery/dv5_service.go

```diff
@@ -319,11 +309,31 @@ func (dvs *DiscV5Service) PublishENR(logger *zap.Logger) {
 		return
 	}
 
-	defer atomic.StoreInt32(&dvs.publishState, publishStateReady)
+	// Acquire publish lock to prevent parallel publishing.
+	ctx, done := context.WithTimeout(dvs.ctx, publishENRTimeout)
+	defer done()
+
+	select {
+	case <-ctx.Done():
+		return
+	case dvs.publishLock <- struct{}{}:
+	}
+	defer func() {
+		// Release lock.
+		<-dvs.publishLock
+	}()
```

Comment:
To avoid potential deadlocks or indefinite waiting, we can make the lock acquisition non-blocking. This ensures that if another routine is already holding the lock, the current routine doesn't block indefinitely but instead logs a warning and skips the current attempt.

```go
ctx, done := context.WithTimeout(dvs.ctx, publishENRTimeout)
defer done()

select {
case dvs.publishLock <- struct{}{}:
    defer func() {
        <-dvs.publishLock // Release the lock
    }()
    
    // the internal logic should go here (the ENR publishing logic)

case <-ctx.Done():
    logger.Warn("skipped ENR publishing: context timed out while waiting for the publish lock")
    return

default:
    logger.Warn("skipped ENR publishing due to another ongoing operation")
    return
}
```

Reply:
@olegshmuelov this would mean that if an ongoing publish is let's say in it's last second (59 out of 60) then the updated record would only be published for 1 second.

in that scenario, the current PR waits 1 second to acquire the lock so that it can continue publishing for 59 more seconds

i updated the comment above the lock to explain the mechanism:
```go
// Acquire publish lock to prevent parallel publishing.
// If there's an ongoing goroutine, it would now start publishing the record updated above,
// and if it's done before the new deadline, this goroutine would pick up where it left off.
```

Reply:
I agree with @moshe-blox here. 
Also on @olegshmuelov suggestion we will rarely wait for the timeout, we will drop into default most times (when lock is acquired).
Waiting is limited to the timeout so I don't see how deadlock can happen currently.

Reply:
Got it, @moshe-blox. That approach makes sense now, and the updated comment clears it up. Looks good!